### ISSUE TITLE
StatefulSet delete behavior changed in v1.11

### ISSUE THREAD

**Issue by rezroo** [kubernetes/kubernetes#68627]
/kind bug

**What happened**:
In k8s v1.11.1 all statefulset pods are deleted at the same time. This is what I observe:
```
stack@master:~$ kubectl get pod -w
NAME           READY     STATUS    RESTARTS   AGE
echoserver-0   0/1       Pending   0          0s
echoserver-0   0/1       Pending   0         0s
echoserver-0   0/1       ContainerCreating   0         0s
echoserver-0   0/1       ContainerCreating   0         0s
echoserver-0   1/1       Running   0         1s
echoserver-1   0/1       Pending   0         0s
echoserver-1   0/1       Pending   0         0s
echoserver-1   0/1       ContainerCreating   0         0s
echoserver-1   0/1       ContainerCreating   0         0s
echoserver-1   1/1       Running   0         1s
echoserver-2   0/1       Pending   0         0s
echoserver-2   0/1       Pending   0         0s
echoserver-2   0/1       ContainerCreating   0         0s
echoserver-2   0/1       ContainerCreating   0         0s
echoserver-2   1/1       Running   0         1s
echoserver-0   1/1       Terminating   0         22s
echoserver-2   1/1       Terminating   0         20s
echoserver-1   1/1       Terminating   0         21s
echoserver-1   0/1       Terminating   0         1m
echoserver-0   0/1       Terminating   0         1m
echoserver-0   0/1       Terminating   0         1m
echoserver-0   0/1       Terminating   0         1m
echoserver-2   0/1       Terminating   0         1m
echoserver-2   0/1       Terminating   0         1m
echoserver-2   0/1       Terminating   0         1m
echoserver-2   0/1       Terminating   0         1m
echoserver-1   0/1       Terminating   0         1m
echoserver-1   0/1       Terminating   0         1m
```

**What you expected to happen**:
In k8s v1.9.3 StatefulSet deletion [adheres to documentation](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#using-statefulsets) and follows `Ordered, graceful deletion and termination`. In v1.9.3 when I delete the statefulset this is what I see:
```
stack@master:~$ kubectl get pod -w
NAME           READY     STATUS    RESTARTS   AGE
echoserver-0   1/1       Running   0          22s
echoserver-1   1/1       Running   0          15s
echoserver-2   1/1       Running   0          9s
echoserver-2   1/1       Terminating   0         29s
echoserver-2   0/1       Terminating   0         30s
echoserver-2   0/1       Terminating   0         38s
echoserver-2   0/1       Terminating   0         38s
echoserver-1   1/1       Terminating   0         44s
echoserver-1   0/1       Terminating   0         45s
echoserver-1   0/1       Terminating   0         55s
echoserver-1   0/1       Terminating   0         55s
echoserver-0   1/1       Terminating   0         1m
echoserver-0   0/1       Terminating   0         1m
echoserver-0   0/1       Terminating   0         1m
echoserver-0   0/1       Terminating   0         1m
```

**How to reproduce it (as minimally and precisely as possible)**:
Create, the delete the following:
```
apiVersion: v1
kind: Service
metadata:
  name: echoserver
  labels:
    app: echoserver
spec:
  ports:
  - port: 80
    targetPort: 8080
  clusterIP: None
  selector:
    app: echoserver
---
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: echoserver
spec:
  serviceName: echoserver
  replicas: 3
  selector:
    matchLabels:
      app: echoserver
  template:
    metadata:
      labels:
        app: echoserver
    spec:
      terminationGracePeriodSeconds: 120
      containers:
      - image: alpine
        name: alpine
        command:
          - sleep
          - '70'
        imagePullPolicy: IfNotPresent
```

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.1", GitCommit:"b1b29978270dc22fecc592ac55d903350454310a", GitTreeState:"clean", BuildDate:"2018-07-17T18:53:20Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.1", GitCommit:"b1b29978270dc22fecc592ac55d903350454310a", GitTreeState:"clean", BuildDate:"2018-07-17T18:43:26Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
NAME="Ubuntu"
VERSION="16.04.2 LTS (Xenial Xerus)"
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

**Comment by krmayankk** [comment]
It seems the default value of pod management policy changed . You can choose parallel and see f that fixes it https://kubernetes.io/docs/tutorials/stateful-application/basic-stateful-set/#pod-management-policy

**Comment by Pingan2017** [comment]
it seems the default deleteoption for `kubectl delete`  changed 
before 1.11, it is  `Foreground` , after it is `Background`
#65908

**Comment by liggitt** [comment]
> before 1.11, it is Foreground , after it is Background

That change was made between 1.11.0 and 1.11.1

**Comment by mortent** [comment]
I think this changed with the removal of reapers in #63979. There used to be a reaper for StatefulSet that took care of orderly and graceful termination from the client, but not termination of pods are done by the garbage collector.
This behavior changed in 1.11.0, but depends upon the client version and not the cluster version. Running a 1.10 client against a 1.11 cluster has the expected behavior.

**Comment by liggitt** [comment]
@smarterclayton @kubernetes/sig-apps-api-reviews for question about statefulset deletion behavior and the podManagementPolicy API field

**Comment by krmayankk** [comment]
Ah I missed this was happening for deletion. And not rolling update . What is the guarantee for deployments ?

**Comment by mortent** [comment]
I have created a PR to update the documentation to reflect the current behavior and guarantees provided by StatefulSets: https://github.com/kubernetes/website/pull/10380

I discussed this with @enisoc and @kow3ns, and there probably is a way to do this using a custom finalizer to prevent the garbage collector from deleting the pods until the StatefulSet controller can scale down to 0. It gets a little more complicated when considering upgrades and rollbacks, but we think it can be solved.

But it is not clear that the benefits of adding this outweighs the cost. It is pretty easy to work around this by simply scaling down the StatefulSet before deleting it. We think that with updated documentation, we can keep the current behavior and revisit it if there demand for a better solution.
